tools/inc/stljobs.h

```diff
@@ -305,14 +315,17 @@ struct output_collecting_pipe {
         writeHandle = create_file(pipeNameBuffer, GENERIC_WRITE | FILE_READ_ATTRIBUTES, 0, &inheritSa, OPEN_EXISTING,
             FILE_FLAG_OVERLAPPED, HANDLE{});
 
+        readEvent         = create_event(nullptr, TRUE, FALSE, nullptr);
+        overlapped.hEvent = readEvent.get();
+
         readIo = tp_io{std::move(readHandle), callback, this, nullptr};
 
         start();
     }
 
     ~output_collecting_pipe() noexcept {
         if (readIo) {
-            if (running.load()) {
+            if (running.exchange(false)) {

```

Comment:
This is broken: if it matters then that means something is trying to destroy output_collecting_pipe while another thread is doing some operation to it; the calling code made the mistake.

Reply:
(It's possible that we might need a WaitForThreadpoolXxx call here rather than this exchange edit)

Reply:
Upon further reflection with @StephanTLavavej I think I understand now: this is to prevent the callback function from calling read_some() again if it gets `ERROR_OPERATION_ABORTED` resulting from the call to `CancelIoEx`.

Perhaps that should be a comment? Or maybe just call stop().

Reply:
A comment along the lines of `// prevent callback() from calling read_some()` would be sufficient. Replacing this with `stop()` would also be reasonable (slightly less efficient, as I understand it, but not in a noticeable way).